### PR TITLE
typing: Session.__init__

### PR DESCRIPTION
**Pull request by blueyed**
Pulled out of https://github.com/pytest-dev/pytest/pull/6491.

### CODE REVIEW

File: src/_pytest/main.py [hunk 1]
```diff
@@ -383,7 +391,7 @@ class Session(nodes.FSCollector):
     # Set on the session by fixtures.pytest_sessionstart.
     _fixturemanager = None  # type: FixtureManager
 
-    def __init__(self, config):
+    def __init__(self, config) -> None:
```
Comment (nicoddemus) [inline src/_pytest/main.py]: Out of curiosity, is considered good practice to annotate the return value of all functions even if they return `None`?

Comment (blueyed) [reply src/_pytest/main.py]: It has to be for the function to be considered to be typed.
I've amended it to also type `config` (the arg) explicitly, although done via base class.

Comment (nicoddemus) [reply src/_pytest/main.py]: I see, thanks makes sense.

Now it brings the next question, is:

```python
def __init__(self, config: Config):
```

enough for it to be considered typed?

Comment (blueyed) [reply src/_pytest/main.py]: Yes (AFAIK), but might fail with strict options etc?!  /cc @bluetech

Comment (blueyed) [reply src/_pytest/main.py]: btw: `reveal_type(var)` is very useful in that regard.

File: src/_pytest/main.py [hunk 2]
```diff
@@ -394,14 +403,16 @@ def __init__(self, config):
         self.trace = config.trace.root.get("collection")
         self._norecursepatterns = config.getini("norecursedirs")
         self.startdir = config.invocation_dir
-        self._initialpaths = frozenset()
+        self._initialpaths = frozenset()  # type: FrozenSet[py.path.local]
+
         # Keep track of any collected nodes in here, so we don't duplicate fixtures
-        self._node_cache = {}
+        self._node_cache = {}  # type: Dict[str, List[Node]]
```
Comment (bluetech) [inline src/_pytest/main.py]: Apparently sometimes the key can be tuple so the type is not exactly correct:

https://github.com/pytest-dev/pytest/blob/ad02f6f8790df4252ba9af69aa76c58eb105a1db/src/_pytest/main.py#L574-L579

I am working on some annotations and will include a fix for this, hopefully tomorrow.

**Pull request merged**